File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/LooseConnectedCellGatherer.java
package org.ligi.gobandroid_hd.logic.cell_gatherer;

import org.ligi.gobandroid_hd.logic.BoardCell;

public class LooseConnectedCellGatherer extends CellGatherer {

    public LooseConnectedCellGatherer(BoardCell root) {
        super(root);
    }

    @Override
    protected void pushWithCheck(BoardCell cell) {
        final boolean unProcessed = processed.add(cell);

        if (cell.isInGroupWith(root)) {
            add(cell);
            if (unProcessed) {
                pushSurroundingWithCheck(cell);
            }
        }

        if (cell.board.isCellFree(cell) && unProcessed) {
            pushSurroundingWithCheck(cell);
        }
    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/LooseConnectedCellGatherer.kt
package org.ligi.gobandroid_hd.logic.cell_gatherer

import org.ligi.gobandroid_hd.logic.StatefulGoBoard
import org.ligi.gobandroid_hd.logic.StatelessBoardCell
import java.util.*

class LooseConnectedCellGatherer(board: StatefulGoBoard, root: StatelessBoardCell) : CellGatherer(board, root) {
    override fun processAndGetFollowup(toProcess: Set<StatelessBoardCell>): HashSet<StatelessBoardCell> {
        val followUp = HashSet<StatelessBoardCell>()
        toProcess.forEach {
            val unProcessed = processed.add(it)

            if (board.areCellsEqual(it, root)) {
                gatheredCells.add(it)
                if (unProcessed) {
                    followUp.addAll(it.neighbors)
                }
            } else if (board.isCellFree(it) && unProcessed) {
                followUp.addAll(it.neighbors)
            }

        }

        return followUp
    }


}
